### ISSUE TITLE
Docker cp does not create intermediate destination directories.

### ISSUE THREAD

**Issue by Zoltu** [docker/docker#20920]
Output of `docker version`:

```
Docker version 1.8.0, build 0d03096
```

Output of `docker info`:

```
Containers: 1
Images: 5
Storage Driver: aufs
 Root Dir: /mnt/sda1/var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 21
 Dirperm1 Supported: true
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 4.1.18-boot2docker
Operating System: Boot2Docker 1.10.2 (TCL 6.4.1); master : 611be10 - Mon Feb 22 22:47:06 UTC 2016
CPUs: 8
Total Memory: 1.955 GiB
Name: boot2docker
ID: PVUD:7SAK:UKCM:SWAM:VO4D:HAVW:WJWE:L7J6:OKOY:EHOE:EOMP:ED3Y
Debug mode (server): true
File Descriptors: 11
Goroutines: 29
System Time: 2016-03-03T17:42:38.217741179Z
EventsListeners: 0
Init SHA1: 
Init Path: /usr/local/bin/docker
Docker Root Dir: /mnt/sda1/var/lib/docker
```

Provide additional environment details (AWS, VirtualBox, physical, etc.):
MacOS X

List the steps to reproduce the issue:
1. Create a docker image with a file at `/app/build/version.txt`
2. Start a container for that image with
`CONTAINER_ID=$(docker create foo:123)`
3. Attempt to copy the file out of the container to a path that doesn't exist with
`docker cp ${CONTAINER_ID}:/app/build/version.txt build/version.txt`
--> `lstat /path/to/working/directory/build: no such file or directory`
4. Try again with just a destination directory with
`docker cp ${CONTAINER_ID}:/app/build/version.txt build/`
--> `no such directory`
5. Create the directory with
`mkdir build`
6. Try again with either command.
--> Notice success.

Describe the results you received:

`no such file or directory`
`no such directory`

Describe the results you expected:

Successful copy of the file.

Provide additional info you think is important:
The documentation [here](https://docs.docker.com/engine/reference/commandline/cp/) is very explicit about the expected behavior in this situation.  In particular, `SRC_PATH` specifies a file and `DEST_PATH` does not exist, so the file should be created at `DEST_PATH`.  The documentation implies that the directory structure leading up to `DEST_PATH` will be created if necessary.

**Comment by thaJeztah** [comment]
Hi @Zoltu, thanks for reporting, I'm actually not sure if this is a bug or the expected behavior (but an incorrect description / incomplete description in the documentation)

I'll defer this one to @jlhawn, who designed/implemented this feature.

**Comment by Zoltu** [comment]
I can accept the current behavior being expected if the documentation is updated to be more clear.  That being said, it really would be nice if docker would create the directories for you when running `cp`, as a quality of life improvement.

**Comment by jlhawn** [comment]
Thanks for the ping @thaJeztah 

Hi @Zoltu!  This is indeed the expected behavior of `docker cp`. The design goal for the utility is for it to behave like your local `cp` command with the `-a` (archive) flag. Try the same set of commands in your example, but using local files only. You'll should see the same behavior.

You're right that the docs could be clearer here:

```
- `SRC_PATH` specifies a file
    - `DEST_PATH` does not exist
        - the file is saved to a file created at `DEST_PATH`
```

It's subtle but the action described is that **a *file* is created at `DEST_PATH`**. It does *not* attempt to create parent directories. This could be cleared up by adding a simple note to all cases in the documentation:

```
`docker cp` does *not* create parent directories for `DEST_PATH`.
```

We could even use your case as an example.

> it really would be nice if docker would create the directories for you when running cp, as a quality of life improvement.

If there is enough demand for it and If it can be shown that there is precedent in other tools like `cp` or `scp` then I would support an option to do this.

**Comment by thaJeztah** [comment]
Thanks @jlhawn!

@Zoltu would you be interested in contributing, and opening a PR to update the documentation? The markdown file used to build that documentation can be found in this repository, here https://github.com/docker/docker/blob/master/docs/reference/commandline/cp.md, and for the man-page; https://github.com/docker/docker/blob/master/man/docker-cp.1.md

If you're interested, we have a contributors guide explaining the contribution workflow https://docs.docker.com/opensource/workflow/make-a-contribution/

Feel free to let us know, otherwise someone else can pick this up :+1:

**Comment by Zoltu** [comment]
If it is a big deal, I can merge those two PRs into one.  Didn't realize it was the same repo until after I submitted them.

**Comment by thaJeztah** [comment]
Oh, please combine them into one, and squash the commits (so that there's a single commit in the PR), that'd be great!

**Comment by likan999** [comment]
I am new to docker. A stupid question, how do I create the intermediate directories in this case? My basically flow is, 
docker create
docker cp ...
docker cp ...
...
docker run

If we can't do it between docker create and docker run, the only way I know is to create my own image just for adding a directory, which is kind of annoying (I need to periodically rebuild the image whenever the upstream one is updated).

**Comment by thaJeztah** [comment]
@likan999 using a Dockerfile, and building your own image is really the best way to go; the whole idea behind Docker is to have reproducible environments (containers); manually creating your container makes that no longer possible.

If you create an auto build image on Docker Hub, you can have the image automatically rebuild if the base image is updated; public images on Docker Hub are free, and you also get one private repository for free

**Comment by asottile** [comment]
I too was trying to figure this out in a situation where dockerfiles weren't really an option.  Here's what I came up with:

```python
def _container_mkdirp(container, path, tmpdir):
    """mkdir -p in a container through continual `docker cp`"""
    path_split = path.split(os.sep)
    for i in range(2, len(path_split) + 1):
        src = os.path.join(tmpdir, *path_split[:i])
        os.makedirs(src, exist_ok=True)
        dest = os.sep.join(path_split[:i])
        cmd = ('docker', 'cp', src + '/.', f'{container.cid}:{dest}')
        subprocess.check_output(cmd)
```

It essentially does this:

```bash
# container=deadbeef
# path=/1/2/3
# tmpdir=/tmp/workdir
mkdir -p /tmp/workdir/1
docker cp /tmp/workdir/1/. deadbeef:/1
mkdir -p /tmp/workdir/1/2
docker cp /tmp/workdir/1/2/. deadbeef:/1/2
mkdir -p /tmp/workdir/1/2/3
docker cp /tmp/workdir/1/2/3/. deadbeef:/1/2/3
```

**Comment by mrtnmark** [comment]
One other solution that builds on @asottile 's answer.
I had a similar situation where I needed to `docker create...` a container, do some operations, `docker cp` the files into the container, and then `docker run`.

I wasn't able to do the mkdir command because the container wasn't currently running yet. A workaround I used was to mount a temporary volume during the `docker create` stage, which in turn created the directory I needed (since volumes create intermediate directories).

```
docker create --name my-container \
-v TMP:/my/long/intermediate/directory \
my-image
```
`docker cp ./mylocalfile.conf my-container:/my/long/intermediate/directory/containerfile.conf`

`docker run my-container`
